jupyter_client/runapp.py

```diff
@@ -0,0 +1,157 @@
+        super(RunApp, self).parse_command_line(argv)
+        self.build_kernel_argv(self.extra_args)
+        self.filenames_to_run = self.extra_args[:]
+
+    @catch_config_error
+    def initialize(self, argv=None):
+        self.log.debug("jupyter run: initialize...")
+        super(RunApp, self).initialize(argv)
+        JupyterConsoleApp.initialize(self)
+        signal.signal(signal.SIGINT, self.handle_sigint)
+        self.init_kernel_info()
+
+    def handle_sigint(self, *args):
+        if self.kernel_manager:
+            self.kernel_manager.interrupt_kernel()
+        else:
+            print("", file=sys.stderr)
+            error("Cannot interrupt kernels we didn't start.\n")
+
+    def init_kernel_info(self):
+        """Wait for a kernel to be ready, and store kernel info"""
+        timeout = self.kernel_timeout
+        tic = time.time()
+        self.kernel_client.hb_channel.unpause()
+        msg_id = self.kernel_client.kernel_info()
+        while True:
+            try:
+                reply = self.kernel_client.get_shell_msg(timeout=1)
+            except queue.Empty:
+                if (time.time() - tic) > timeout:
+                    raise RuntimeError("Kernel didn't respond to kernel_info_request")
+            else:
+                if reply['parent_header'].get('msg_id') == msg_id:
+                    self.kernel_info = reply['content']
+                    return
+
+    def start(self):
+        self.log.debug("jupyter run: starting...")
+        super(RunApp, self).start()
+        for filename in self.filenames_to_run:
+            self.log.debug("jupyter run: executing `%s`" % filename)
+            with open(filename) as fp:
+                cell = fp.read()
+            return_code = self.execute_printing_output(cell)
+            if return_code:
+                raise Exception("jupyter-run error running '%s'" % filename)
+
+    def execute_printing_output(self, cell):
+        """
+        Run a cell on a KernelClient
+        Any output from the cell will be displayed.
+        """
+        msg_id = self.kernel_client.execute(cell)
+        return_code = 0
+        while True:
+            try:
+                msg = self.kernel_client.get_iopub_msg(timeout=OUTPUT_TIMEOUT)
+            except queue.Empty:
+                raise TimeoutError("Timeout waiting for kernel output")
+    
+            if msg['parent_header'].get('msg_id') != msg_id:
+                continue
+            msg_type = msg['header']['msg_type']
+            content = msg['content']
+            if msg_type == 'status':
+                if content['execution_state'] == 'idle':
+                    # idle means output is done
+                    break
+            elif msg_type == 'stream':
+                stream = getattr(sys, content['name'])
+                stream.write(content['text'])
+                # TODO: remove this when all kernels use error msg_type:
+                if content['name'] == "stderr":
```

Comment:
No, really, we shouldn't do this. It breaks kernels that do signal errors correctly to make a vague affordance to kernels that don't.

I suspect that when we correctly check the `execute_reply` message in addition to this, we'll find that most kernels do work correctly.